File: src/tablet/otd/configLoader.ts
/** Loader and matcher for vendored OpenTabletDriver configurations. */

import { promises as fs } from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Default location of the OTD JSON configs. The Node and Python ports
 * share the same vendored copy under <repo-root>/configs to avoid drift.
 */
export const CONFIGS_ROOT = path.resolve(__dirname, "..", "..", "..", "configs");

export interface DigitizerIdentifier {
  vendorId: number;
  productId: number;
  inputReportLength: number | null;
  outputReportLength: number | null;
  reportParser: string;
  deviceStrings: Record<string, string>;
  initializationStrings: number[];
  featureInitReports: Uint8Array[];
  outputInitReports: Uint8Array[];
}

export interface TabletSpecifications {
  digitizerWidthMm: number;
  digitizerHeightMm: number;
  digitizerMaxX: number;
  digitizerMaxY: number;
  penMaxPressure: number;
  penButtonCount: number;
  auxButtonCount: number;
}

export class TabletConfiguration {
  constructor(
    public name: string,
    public sourcePath: string,
    public specifications: TabletSpecifications,
    public digitizerIdentifiers: DigitizerIdentifier[],
    public libinputOverride = false,
  ) {}

  get manufacturer(): string { return this.name.split(" ", 1)[0]; }
  get model(): string {
    const idx = this.name.indexOf(" ");
    return idx === -1 ? this.name : this.name.slice(idx + 1);
  }
}

function parseSpecs(raw: any): TabletSpecifications {
  const digi = raw?.Digitizer ?? {};
  const pen = raw?.Pen ?? {};
  const aux = raw?.AuxiliaryButtons ?? {};
  return {
    digitizerWidthMm: Number(digi.Width ?? 0) || 0,
    digitizerHeightMm: Number(digi.Height ?? 0) || 0,
    digitizerMaxX: Number(digi.MaxX ?? 0) || 0,
    digitizerMaxY: Number(digi.MaxY ?? 0) || 0,
    penMaxPressure: Number(pen.MaxPressure ?? 0) || 0,
    penButtonCount: Number(pen.ButtonCount ?? 0) || 0,
    auxButtonCount: Number(aux.ButtonCount ?? 0) || 0,
  };
}

/** OTD serializes byte arrays in JSON as base64 strings or int arrays. */
function decodeReport(value: unknown): Uint8Array {
  if (typeof value === "string") {
    return new Uint8Array(Buffer.from(value, "base64"));
  }
  if (Array.isArray(value)) {
    return new Uint8Array(value.map((b) => Number(b) & 0xFF));
  }
  return new Uint8Array();
}

function requireField<T>(obj: any, key: string): T {
  if (obj == null || !(key in obj)) {
    throw new Error(`missing required field: ${key}`);
  }
  return obj[key] as T;
}

function parseIdentifier(raw: any): DigitizerIdentifier {
  const deviceStrings: Record<string, string> = {};
  for (const [k, v] of Object.entries(raw.DeviceStrings ?? {})) {
    deviceStrings[String(k)] = String(v);
  }
  const initStrings = (raw.InitializationStrings ?? []).map(
    (i: number) => Number(i) & 0xFF,
  );
  const featureInit = (raw.FeatureInitReport ?? [])
    .map(decodeReport).filter((b: Uint8Array) => b.length > 0);
  const outputInit = (raw.OutputInitReport ?? [])
    .map(decodeReport).filter((b: Uint8Array) => b.length > 0);
  return {
    vendorId: Number(requireField(raw, "VendorID")),
    productId: Number(requireField(raw, "ProductID")),
    inputReportLength: raw.InputReportLength ?? null,
    outputReportLength: raw.OutputReportLength ?? null,
    reportParser: String(requireField(raw, "ReportParser")),
    deviceStrings,
    initializationStrings: initStrings,
    featureInitReports: featureInit,
    outputInitReports: outputInit,
  };
}

export async function loadConfigFile(filePath: string): Promise<TabletConfiguration> {
  const raw = JSON.parse(await fs.readFile(filePath, "utf-8"));
  const attrs = raw.Attributes ?? {};
  return new TabletConfiguration(
    String(requireField(raw, "Name")),
    filePath,
    parseSpecs(raw.Specifications ?? {}),
    (raw.DigitizerIdentifiers ?? []).map(parseIdentifier),
    String(attrs.libinputoverride ?? "0") !== "0",
  );
}

async function* walkJson(dir: string): AsyncGenerator<string> {
  const entries = await fs.readdir(dir, { withFileTypes: true });
  for (const entry of entries.sort((a, b) => a.name.localeCompare(b.name))) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      yield* walkJson(full);
    } else if (entry.isFile() && entry.name.endsWith(".json")) {
      yield full;
    }
  }
}

export async function loadAllConfigs(root: string = CONFIGS_ROOT): Promise<TabletConfiguration[]> {
  const out: TabletConfiguration[] = [];
  for await (const file of walkJson(root)) {
    try {
      out.push(await loadConfigFile(file));
    } catch {
      continue;
    }
  }
  return out;
}

export interface ConfigMatch {
  config: TabletConfiguration;
  identifier: DigitizerIdentifier;
}

export class ConfigIndex {
  private byVidPid = new Map<string, ConfigMatch[]>();

  constructor(configs: Iterable<TabletConfiguration>) {
    for (const cfg of configs) {
      for (const ident of cfg.digitizerIdentifiers) {
        const key = `${ident.vendorId}:${ident.productId}`;
        const list = this.byVidPid.get(key) ?? [];
        list.push({ config: cfg, identifier: ident });
        this.byVidPid.set(key, list);
      }
    }
  }

  static async fromVendored(): Promise<ConfigIndex> {
    return new ConfigIndex(await loadAllConfigs());
  }

  find(vendorId: number, productId: number,
       inputReportLength: number | null = null): ConfigMatch[] {
    const candidates = [...(this.byVidPid.get(`${vendorId}:${productId}`) ?? [])];
    if (inputReportLength !== null) {
      const narrowed = candidates.filter((m) =>
        m.identifier.inputReportLength === null ||
        m.identifier.inputReportLength === inputReportLength);
      if (narrowed.length > 0) return narrowed;
    }
    return candidates;
  }

  allVidPidPairs(): Array<[number, number]> {
    return [...this.byVidPid.keys()]
      .map((k) => k.split(":").map(Number) as [number, number])
      .sort((a, b) => (a[0] - b[0]) || (a[1] - b[1]));
  }
}

File: python/sketchatone/tablet/otd/config_loader.py
"""Loader and matcher for vendored OpenTabletDriver configurations."""

from __future__ import annotations

import base64
import json
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, Iterable, List, Optional, Tuple


# Bundled alongside the module so the JSONs travel with any install of the
# ``sketchatone`` package (source checkout, wheel, or .deb) without the
# packaging scripts needing to know about them.
CONFIGS_ROOT = Path(__file__).resolve().parent / "configs"


@dataclass(frozen=True)
class DigitizerIdentifier:
    """One entry from a config's ``DigitizerIdentifiers`` array."""
    vendor_id: int
    product_id: int
    input_report_length: Optional[int]
    output_report_length: Optional[int]
    report_parser: str
    device_strings: Dict[str, str] = field(default_factory=dict)
    initialization_strings: Tuple[int, ...] = ()
    feature_init_reports: Tuple[bytes, ...] = ()
    output_init_reports: Tuple[bytes, ...] = ()


@dataclass(frozen=True)
class TabletSpecifications:
    """Subset of OTD Specifications we actually use for normalization."""
    digitizer_width_mm: float = 0.0
    digitizer_height_mm: float = 0.0
    digitizer_max_x: int = 0
    digitizer_max_y: int = 0
    pen_max_pressure: int = 0
    pen_button_count: int = 0
    aux_button_count: int = 0


@dataclass
class TabletConfiguration:
    """One vendored OTD configuration file."""
    name: str
    source_path: Path
    specifications: TabletSpecifications
    digitizer_identifiers: List[DigitizerIdentifier]
    libinput_override: bool = False

    @property
    def manufacturer(self) -> str:
        return self.name.split(" ", 1)[0]

    @property
    def model(self) -> str:
        parts = self.name.split(" ", 1)
        return parts[1] if len(parts) > 1 else self.name


def _parse_specs(raw: dict) -> TabletSpecifications:
    digi = raw.get("Digitizer") or {}
    pen = raw.get("Pen") or {}
    aux = raw.get("AuxiliaryButtons") or {}
    return TabletSpecifications(
        digitizer_width_mm=float(digi.get("Width", 0) or 0),
        digitizer_height_mm=float(digi.get("Height", 0) or 0),
        digitizer_max_x=int(digi.get("MaxX", 0) or 0),
        digitizer_max_y=int(digi.get("MaxY", 0) or 0),
        pen_max_pressure=int(pen.get("MaxPressure", 0) or 0),
        pen_button_count=int(pen.get("ButtonCount", 0) or 0),
        aux_button_count=int(aux.get("ButtonCount", 0) or 0),
    )


def _decode_report(value) -> bytes:
    """OTD serializes byte arrays in JSON as base64 strings."""
    if isinstance(value, str):
        return base64.b64decode(value)
    if isinstance(value, list):
        return bytes(int(b) & 0xFF for b in value)
    return b""


def _parse_identifier(raw: dict) -> DigitizerIdentifier:
    device_strings = raw.get("DeviceStrings") or {}
    init_strings = tuple(int(i) & 0xFF for i in (raw.get("InitializationStrings") or []))
    feature_init = tuple(_decode_report(r) for r in (raw.get("FeatureInitReport") or []))
    output_init = tuple(_decode_report(r) for r in (raw.get("OutputInitReport") or []))
    return DigitizerIdentifier(
        vendor_id=int(raw["VendorID"]),
        product_id=int(raw["ProductID"]),
        input_report_length=raw.get("InputReportLength"),
        output_report_length=raw.get("OutputReportLength"),
        report_parser=raw["ReportParser"],
        device_strings={str(k): str(v) for k, v in device_strings.items()},
        initialization_strings=init_strings,
        feature_init_reports=tuple(r for r in feature_init if r),
        output_init_reports=tuple(r for r in output_init if r),
    )


def load_config_file(path: Path) -> TabletConfiguration:
    raw = json.loads(path.read_text(encoding="utf-8"))
    return TabletConfiguration(
        name=raw["Name"],
        source_path=path,
        specifications=_parse_specs(raw.get("Specifications") or {}),
        digitizer_identifiers=[_parse_identifier(d) for d in raw.get("DigitizerIdentifiers", [])],
        libinput_override=str((raw.get("Attributes") or {}).get("libinputoverride", "0")) != "0",
    )


def load_all_configs(root: Optional[Path] = None) -> List[TabletConfiguration]:
    base = root or CONFIGS_ROOT
    configs: List[TabletConfiguration] = []
    skipped: List[Tuple[Path, str]] = []
    if not base.is_dir():
        print(
            f"[OTD] Warning: tablet config directory not found: {base}",
            file=sys.stderr,
        )
        return configs
    for path in sorted(base.rglob("*.json")):
        try:
            configs.append(load_config_file(path))
        except (KeyError, ValueError, json.JSONDecodeError) as e:
            skipped.append((path, type(e).__name__))
            continue
    if not configs:
        print(
            f"[OTD] Warning: no tablet configs loaded from {base} "
            f"({len(skipped)} file(s) skipped)",
            file=sys.stderr,
        )
    return configs


@dataclass
class ConfigMatch:
    """A configuration matched against a connected HID device."""
    config: TabletConfiguration
    identifier: DigitizerIdentifier


class ConfigIndex:
    """In-memory index for fast lookup by (VID, PID) and optionally report length."""

    def __init__(self, configs: Iterable[TabletConfiguration]):
        self._by_vid_pid: Dict[Tuple[int, int], List[ConfigMatch]] = {}
        for cfg in configs:
            for ident in cfg.digitizer_identifiers:
                self._by_vid_pid.setdefault((ident.vendor_id, ident.product_id), []).append(
                    ConfigMatch(config=cfg, identifier=ident)
                )

    @classmethod
    def from_vendored(cls) -> "ConfigIndex":
        return cls(load_all_configs())

    def find(self, vendor_id: int, product_id: int,
             input_report_length: Optional[int] = None) -> List[ConfigMatch]:
        candidates = list(self._by_vid_pid.get((vendor_id, product_id), []))
        if input_report_length is not None:
            narrowed = [
                m for m in candidates
                if m.identifier.input_report_length in (None, input_report_length)
            ]
            if narrowed:
                return narrowed
        return candidates

    def all_vid_pid_pairs(self) -> List[Tuple[int, int]]:
        return sorted(self._by_vid_pid)
